# 习题 5.4

![习题 5.4](https://tva1.sinaimg.cn/large/008i3skNly1grbv1zgy77j316008ejut.jpg)

In [1]:
data <- read.csv("ex_5_4.csv")[-1]
attach(data)
data

group x1  x2  x3  x4  x5 x6   x7 
1  1     6.6  39 1.0 6.0  6 0.12  20
2  1     6.6  39 1.0 6.0 12 0.12  20
3  1     6.1  47 1.0 6.0  6 0.08  12
4  1     6.1  47 1.0 6.0 12 0.08  12
5  1     8.4  32 2.0 7.5 19 0.35  75
6  1     7.2   6 1.0 7.0 28 0.30  30
7  1     8.4 113 3.5 6.0 18 0.15  75
8  1     7.5  52 1.0 6.0 12 0.16  40
9  1     7.5  52 3.5 7.5  6 0.16  40
10 1     8.3 113 0.0 7.5 35 0.12 180
11 1     7.8 172 1.0 3.5 14 0.21  45
12 1     7.8 172 1.5 3.0 15 0.21  45
13 2     8.4  32 1.0 5.0  4 0.35  75
14 2     8.4  32 2.0 9.0 10 0.35  75
15 2     8.4  32 2.5 4.0 10 0.35  75
16 2     6.3  11 4.5 7.5  3 0.20  15
17 2     7.0   8 4.5 4.5  9 0.25  30
18 2     7.0   8 6.0 7.5  4 0.25  30
19 2     7.0   8 1.5 6.0  1 0.25  30
20 2     8.3 161 1.5 4.0  4 0.08  70
21 2     8.3 161 0.5 2.5  1 0.08  70
22 2     7.2   6 3.5 4.0 12 0.30  30
23 2     7.2   6 1.0 3.0  3 0.30  30
24 2     7.2   6 1.0 6.0  5 0.30  30
25 2     5.5   6 2.5 3.0  7 0.18  18
26 2     8.4 113 3.5 4.5  6 0.15  75
27 2     8.4 113 3.5 4.5  8 0.15  75
28 2     7.5  52 1.0 6.0  6 0.16  40
29 2     7.5  52 1.0 7.5  8 0.16  40
30 2     8.3  97 0.0 6.0  5 0.15 180
31 2     8.3  97 2.5 6.0  5 0.15 180
32 2     8.3  89 0.0 6.0 10 0.16 180
33 2     8.3  56 1.5 6.0 13 0.25 180
34 2     7.8 172 1.0 3.5  6 0.21  45
35 2     7.8 233 1.0 4.5  6 0.18  45

通过 MASS 包的 lda 进行判别分析：

In [2]:
library(MASS)

## 先验概率按比例分配

In [3]:
l = lda(group ~ ., data=data); l

Call:
lda(group ~ ., data = data)

Prior probabilities of groups:
        1         2 
0.3428571 0.6571429 

Group means:
        x1       x2       x3      x4        x5        x6       x7
1 7.358333 73.66667 1.458333 6.00000 15.250000 0.1716667 49.50000
2 7.686957 67.43478 2.043478 5.23913  6.347826 0.2156522 70.34783

Coefficients of linear discriminants:
             LD1
x1 -1.747293e-01
x2  1.416021e-05
x3  2.072826e-01
x4 -2.052228e-01
x5 -1.935168e-01
x6  8.395266e+00
x7  1.917129e-02

输出结果中 (ref https://cloud.tencent.com/developer/article/1553504 )：

- Call 表示调用方法；
- Prior probabilities of groups 表示先验概率；
- Group means 表示每一类样本的均值；
- Coefficients of linear discriminants 表示线性判别系数；

In [4]:
# 输出线性判别函数
cat("W =", paste(round(l$scaling, 8), "*x_", 1:7, " +", sep=""))

W = -0.17472934*x_1 + 1.416e-05*x_2 + 0.20728259*x_3 + -0.20522277*x_4 + -0.1935168*x_5 + 8.395266*x_6 + 0.01917129*x_7 +

In [5]:
pred <- predict(l)
pred.tab <- data.frame(pred, data.group=group); pred.tab

class posterior.1  posterior.2  LD1        data.group
1  1     0.6859662059 3.140338e-01 -0.9541788 1         
2  1     0.9807303897 1.926961e-02 -2.1152796 1         
3  1     0.8665260167 1.334740e-01 -1.3558819 1         
4  1     0.9934324560 6.567544e-03 -2.5169826 1         
5  1     0.6533234780 3.466765e-01 -0.8997285 1         
6  1     0.9998065233 1.934767e-04 -3.8192153 1         
7  1     0.9517049526 4.829505e-02 -1.7653599 1         
8  1     0.9172384730 8.276153e-02 -1.5531155 1         
9  2     0.2119047770 7.880952e-01 -0.1816424 1         
10 1     0.9999488458 5.115424e-05 -4.3098682 1         
11 1     0.6906278386 3.093722e-01 -0.9621920 1         
12 1     0.6832009800 3.167990e-01 -0.9494561 1         
13 2     0.0003137630 9.996862e-01  2.3087978 2         
14 2     0.0371741982 9.628258e-01  0.5340885 2         
15 2     0.0018008368 9.981992e-01  1.6638437 2         
16 2     0.0258888529 9.741111e-01  0.6718136 2         
17 2     0.0233176914 9.766823e-01  0.7113608 2         
18 2     0.0039412644 9.960587e-01  1.3742004 2         
19 2     0.0044431240 9.955569e-01  1.3298132 2         
20 2     0.0724237508 9.275762e-01  0.2743832 2         
21 2     0.0121654340 9.878346e-01  0.9554851 2         
22 2     0.0511593102 9.488407e-01  0.4109283 2         
23 2     0.0011189212 9.988811e-01  1.8395958 2         
24 2     0.0167026449 9.832974e-01  0.8368939 2         
25 2     0.0479575131 9.520425e-01  0.4360049 2         
26 2     0.0155093991 9.844906e-01  0.8646758 2         
27 2     0.0430578686 9.569421e-01  0.4776422 2         
28 2     0.3223405288 6.776595e-01 -0.3920147 2         
29 1     0.7578938074 2.421062e-01 -1.0868825 2         
30 2     0.0006240626 9.993759e-01  2.0551013 2         
31 2     0.0001531707 9.998468e-01  2.5733078 2         
32 2     0.0068117313 9.931883e-01  1.1713567 2         
33 2     0.0018353157 9.981647e-01  1.6568369 2         
34 2     0.0324563440 9.675437e-01  0.5859424 2         
35 2     0.1036058164 8.963942e-01  0.1297254 2

输出结果分别为分类结果和后验概率。在后面加上了数据的真实分类情况作为比较。

下面查看误判情况：

In [6]:
table(group, pred$class)

     
group  1  2
    1 11  1
    2  1 22

In [7]:
bad <- pred.tab[pred$class != data$group,]; bad

class posterior.1 posterior.2 LD1        data.group
9  2     0.2119048   0.7880952   -0.1816424 1         
29 1     0.7578938   0.2421062   -1.0868825 2

In [8]:
length(bad$class) / length(pred$class)

[1] 0.05714286

有两个样本误判：

- 9 号样本误判到 $G_2$；
- 29 号样本误判到 $G_1$.

误判率为: $5.71\%$

交叉确认误判率 (ref https://zhuanlan.zhihu.com/p/23965433 )：

In [9]:
ll <- lda(group ~ ., data=data, CV=TRUE)

In [10]:
table(group, ll$class)

     
group  1  2
    1  8  4
    2  2 21

In [11]:
cv.bad <- pred.tab[ll$class != data$group,]; cv.bad

class posterior.1 posterior.2 LD1        data.group
5  1     0.6533235   0.3466765   -0.8997285 1         
9  2     0.2119048   0.7880952   -0.1816424 1         
11 1     0.6906278   0.3093722   -0.9621920 1         
12 1     0.6832010   0.3167990   -0.9494561 1         
29 1     0.7578938   0.2421062   -1.0868825 2         
35 2     0.1036058   0.8963942    0.1297254 2

In [12]:
length(cv.bad$class) / length(pred$class)

[1] 0.1714286

共有 6 个样本误判，误判率 $17.14\%$.

## 先验概率相等

假设 $\Sigma_1=\Sigma_2$， 先验概率按相同的条件下，还是用 lda 函数来做，需要多传一个参数指定先验概率。

In [13]:
le <- lda(group ~ ., data = data, prior = c(1, 1) / 2)
print(le)

Call:
lda(group ~ ., data = data, prior = c(1, 1)/2)

Prior probabilities of groups:
  1   2 
0.5 0.5 

Group means:
        x1       x2       x3      x4        x5        x6       x7
1 7.358333 73.66667 1.458333 6.00000 15.250000 0.1716667 49.50000
2 7.686957 67.43478 2.043478 5.23913  6.347826 0.2156522 70.34783

Coefficients of linear discriminants:
             LD1
x1 -1.747293e-01
x2  1.416021e-05
x3  2.072826e-01
x4 -2.052228e-01
x5 -1.935168e-01
x6  8.395266e+00
x7  1.917129e-02


In [14]:
# 输出线性判别函数
cat("W =", paste(round(le$scaling, 8), "*x_", 1:7, " +", sep=""))

W = -0.17472934*x_1 + 1.416e-05*x_2 + 0.20728259*x_3 + -0.20522277*x_4 + -0.1935168*x_5 + 8.395266*x_6 + 0.01917129*x_7 +

In [15]:
pred_e <- predict(le)
pred_e.tab <- data.frame(pred_e, data.group=group); pred.tab

class posterior.1  posterior.2  LD1        data.group
1  1     0.6859662059 3.140338e-01 -0.9541788 1         
2  1     0.9807303897 1.926961e-02 -2.1152796 1         
3  1     0.8665260167 1.334740e-01 -1.3558819 1         
4  1     0.9934324560 6.567544e-03 -2.5169826 1         
5  1     0.6533234780 3.466765e-01 -0.8997285 1         
6  1     0.9998065233 1.934767e-04 -3.8192153 1         
7  1     0.9517049526 4.829505e-02 -1.7653599 1         
8  1     0.9172384730 8.276153e-02 -1.5531155 1         
9  2     0.2119047770 7.880952e-01 -0.1816424 1         
10 1     0.9999488458 5.115424e-05 -4.3098682 1         
11 1     0.6906278386 3.093722e-01 -0.9621920 1         
12 1     0.6832009800 3.167990e-01 -0.9494561 1         
13 2     0.0003137630 9.996862e-01  2.3087978 2         
14 2     0.0371741982 9.628258e-01  0.5340885 2         
15 2     0.0018008368 9.981992e-01  1.6638437 2         
16 2     0.0258888529 9.741111e-01  0.6718136 2         
17 2     0.0233176914 9.766823e-01  0.7113608 2         
18 2     0.0039412644 9.960587e-01  1.3742004 2         
19 2     0.0044431240 9.955569e-01  1.3298132 2         
20 2     0.0724237508 9.275762e-01  0.2743832 2         
21 2     0.0121654340 9.878346e-01  0.9554851 2         
22 2     0.0511593102 9.488407e-01  0.4109283 2         
23 2     0.0011189212 9.988811e-01  1.8395958 2         
24 2     0.0167026449 9.832974e-01  0.8368939 2         
25 2     0.0479575131 9.520425e-01  0.4360049 2         
26 2     0.0155093991 9.844906e-01  0.8646758 2         
27 2     0.0430578686 9.569421e-01  0.4776422 2         
28 2     0.3223405288 6.776595e-01 -0.3920147 2         
29 1     0.7578938074 2.421062e-01 -1.0868825 2         
30 2     0.0006240626 9.993759e-01  2.0551013 2         
31 2     0.0001531707 9.998468e-01  2.5733078 2         
32 2     0.0068117313 9.931883e-01  1.1713567 2         
33 2     0.0018353157 9.981647e-01  1.6568369 2         
34 2     0.0324563440 9.675437e-01  0.5859424 2         
35 2     0.1036058164 8.963942e-01  0.1297254 2

误判情况：

In [16]:
table(group, pred_e$class)

     
group  1  2
    1 11  1
    2  1 22

In [17]:
bad_e <- pred_e.tab[pred_e$class != data$group,]; bad_e

class posterior.1 posterior.2 LD1        data.group
9  2     0.3400897   0.6599103    0.2444662 1         
29 1     0.8571422   0.1428578   -0.6607739 2

In [18]:
# 误判率
length(bad_e$class) / length(pred_e$class)

[1] 0.05714286

交叉确认：

In [19]:
lel <- lda(group ~ ., data=data,  prior = c(1, 1) / 2, CV=TRUE)
table(group, lel$class)

     
group  1  2
    1 11  1
    2  3 20

In [20]:
cv.bad_e <- pred_e.tab[lel$class != data$group,]; cv.bad_e

class posterior.1 posterior.2 LD1         data.group
9  2     0.3400897   0.6599103    0.24446615 1         
28 2     0.4769042   0.5230958    0.03409383 2         
29 1     0.8571422   0.1428578   -0.66077393 2         
35 2     0.1813542   0.8186458    0.55583396 2

In [21]:
# 误判率
length(cv.bad_e$class) / length(pred_e$class)

[1] 0.1142857

这里得到了误判率 $11.43\%$.